src/librustc_mir/interpret/visitor.rs

```diff
@@ -0,0 +1,125 @@
+//! Visitor for a run-time value with a given layout: Traverse enums, structs and other compound
+//! types until we arrive at the leaves, with custom handling for primitive types.
+
+use std::fmt;
+
+use rustc::ty::layout::{self, TyLayout};
+use rustc::ty;
+use rustc::mir::interpret::{
+    EvalResult,
+};
+
+use super::{
+    Machine, EvalContext,
+};
+
+// How to traverse a value and what to do when we are at the leaves.
+// In the future, we might want to turn this into two traits, but so far the
+// only implementations we have couldn't share any code anyway.
+pub trait ValueVisitor<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>>: fmt::Debug {
+    // Get this value's layout.
+    fn layout(&self) -> TyLayout<'tcx>;
+
+    // Downcast functions.  These change the value as a side-effect.
+    fn downcast_enum(&mut self, ectx: &EvalContext<'a, 'mir, 'tcx, M>)
+        -> EvalResult<'tcx>;
+    fn downcast_dyn_trait(&mut self, ectx: &EvalContext<'a, 'mir, 'tcx, M>)
```

Comment:
I dislike the `self` mutation. It does not feel right. I think these downcast functions should be recursing via `visit_value`.

Reply:
`downcast_dyn_trait` is gone in a later commit.

For `downcast_enum` I will do this, but it actually makes the interface larger overall (i.e., one more method to implement).

Reply:
Oh and then there is an extra hook I need to keep the nice error reporting... of the validity check :/

Reply:
Okay I did something. Thanks for making me do it, it is now better than it was before. :)  Let me know what you think.